Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp

```diff
@@ -0,0 +1,503 @@
+    return FileSystem::pathByAppendingComponent(rootDirectory, key.hashAsString());
+}
+
+HashSet<WebCore::ClientOrigin> CacheStorageManager::originsOfCacheStorageData(const String& rootDirectory)
+{
+    HashSet<WebCore::ClientOrigin> result;
+    for (auto& originName : FileSystem::listDirectory(rootDirectory)) {
+        auto originFile = FileSystem::pathByAppendingComponents(rootDirectory, { originName, originFileName });
+        if (auto origin = readOriginFromFile(originFile))
+            result.add(*origin);
+    }
+
+    return result;
+}
+
+static uint64_t getDirectorySize(const String& originDirectory)
+{
+    uint64_t directorySize = 0;
+    Deque<String> paths;
+    paths.append(originDirectory);
+    while (!paths.isEmpty()) {
+        auto path = paths.takeFirst();
+        if (FileSystem::fileType(path) == FileSystem::FileType::Directory) {
+            auto fileNames = FileSystem::listDirectory(path);
+            for (auto& fileName : fileNames) {
+                // Files in /Blobs directory are hard link.
+                if (fileName == "Blobs"_s)
+                    continue;
+                paths.append(FileSystem::pathByAppendingComponent(path, fileName));
+            }
+            continue;
+        }
+        directorySize += FileSystem::fileSize(path).value_or(0);
+    }
+    return directorySize;
+}
+
+uint64_t CacheStorageManager::cacheStorageSize(const String& originDirectory)
+{
+    if (auto sizeFromFile = readSizeFile(originDirectory))
+        return *sizeFromFile;
+
+    return getDirectorySize(originDirectory);
+}
+
+bool CacheStorageManager::hasCacheList(const String& cacheStorageDirectory)
+{
+    return FileSystem::fileExists(FileSystem::pathByAppendingComponent(cacheStorageDirectory, cachesListFileName));
+}
+
+void CacheStorageManager::makeDirty()
+{
+    m_updateCounter = nextUpdateNumber();
+}
+
+CacheStorageManager::CacheStorageManager(const String& path, FileSystem::Salt salt, CacheStorageRegistry& registry, const WebCore::ClientOrigin& origin, QuotaCheckFunction&& quotaCheckFunction, Ref<WorkQueue>&& queue)
+    : m_updateCounter(nextUpdateNumber())
+    , m_path(path)
+    , m_salt(salt)
+    , m_registry(registry)
+    , m_quotaCheckFunction(WTFMove(quotaCheckFunction))
+    , m_queue(WTFMove(queue))
+{
+    if (!m_path.isEmpty()) {
+        auto originFile = FileSystem::pathByAppendingComponent(m_path, originFileName);
+        writeOriginToFile(originFile, origin);
+    }
+}
+
+CacheStorageManager::~CacheStorageManager()
+{
+    for (auto& request : m_pendingSpaceRequests)
+        request.second(false);
+
+    for (auto& cache : m_caches)
+        m_registry.unregisterCache(cache->identifier());
+
+    for (auto& identifier : m_removedCaches.keys())
+        m_registry.unregisterCache(identifier);
+}
+
+bool CacheStorageManager::initializeCaches()
+{
+    if (m_isInitialized)
+        return true;
+
+    auto cachesList = readCachesList(m_path);
```

Comment:
Should it be read asynchronously?

Reply:
It would mean one client origin could delay another client origin since they would all use the same thread, correct?
If so, we should do this. A follow-up is probably fine but then let's keep track of this work, as well as the optimization work.

Reply:
Yes, I will add a comment here and file a bug.

Reply:
Filed https://bugs.webkit.org/show_bug.cgi?id=249593.